File: src/main/java/org/cobbzilla/util/system/Command.java
package org.cobbzilla.util.system;

import lombok.experimental.Accessors;
import org.apache.commons.exec.CommandLine;
import org.apache.commons.lang3.ArrayUtils;
import org.cobbzilla.util.collection.SingletonList;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static org.cobbzilla.util.daemon.ZillaRuntime.empty;
import static org.cobbzilla.util.io.FileUtil.abs;
import static org.cobbzilla.util.string.StringUtil.UTF8cs;

@Accessors(chain=true)
public class Command {

    public static final List<Integer> DEFAULT_EXIT_VALUES = new SingletonList<>(0);
    public static final int[] DEFAULT_EXIT_VALUES_INT = { 0 };

    private CommandLine commandLine;
    private String input;
    private byte[] rawInput;
    private InputStream stdin;
    private File dir;
    private Map<String, String> env;
    private List<Integer> exitValues = DEFAULT_EXIT_VALUES;

    private boolean copyToStandard = false;

    private OutputStream out;

    public Command() {
    }

    public boolean hasOut () { return out != null; }

    private OutputStream err;
    public boolean hasErr () { return err != null; }

    public Command(CommandLine commandLine) { this.commandLine = commandLine; }
    public Command(String command) { this(CommandLine.parse(command)); }
    public Command(File executable) { this(abs(executable)); }

    public boolean hasDir () { return dir != null; }
    public boolean hasInput () { return !empty(input) || !empty(rawInput) || stdin != null; }
    public InputStream getInputStream () {
        if (!hasInput()) return null;
        if (stdin != null) return stdin;
        if (rawInput != null) return new ByteArrayInputStream(rawInput);
        return new ByteArrayInputStream(input.getBytes(UTF8cs));
    }

    public int[] getExitValues () {
        return exitValues == DEFAULT_EXIT_VALUES
                ? DEFAULT_EXIT_VALUES_INT
                : ArrayUtils.toPrimitive(exitValues.toArray(new Integer[exitValues.size()]));
    }

    public Command setExitValues (List<Integer> values) { this.exitValues = values; return this; }

    public Command setExitValues (int[] values) {
        exitValues = new ArrayList<>(values.length);
        for (int v : values) exitValues.add(v);
        return this;
    }

    public CommandLine getCommandLine() {
        return this.commandLine;
    }

    public String getInput() {
        return this.input;
    }

    public byte[] getRawInput() {
        return this.rawInput;
    }

    public InputStream getStdin() {
        return this.stdin;
    }

    public File getDir() {
        return this.dir;
    }

    public Map<String, String> getEnv() {
        return this.env;
    }

    public boolean isCopyToStandard() {
        return this.copyToStandard;
    }

    public OutputStream getOut() {
        return this.out;
    }

    public OutputStream getErr() {
        return this.err;
    }

    public Command setCommandLine(CommandLine commandLine) {
        this.commandLine = commandLine;
        return this;
    }

    public Command setInput(String input) {
        this.input = input;
        return this;
    }

    public Command setRawInput(byte[] rawInput) {
        this.rawInput = rawInput;
        return this;
    }

    public Command setStdin(InputStream stdin) {
        this.stdin = stdin;
        return this;
    }

    public Command setDir(File dir) {
        this.dir = dir;
        return this;
    }

    public Command setEnv(Map<String, String> env) {
        this.env = env;
        return this;
    }

    public Command setCopyToStandard(boolean copyToStandard) {
        this.copyToStandard = copyToStandard;
        return this;
    }

    public Command setOut(OutputStream out) {
        this.out = out;
        return this;
    }

    public Command setErr(OutputStream err) {
        this.err = err;
        return this;
    }
}

File: src/main/java/org/cobbzilla/util/system/Command.kt
package org.cobbzilla.util.system

import lombok.experimental.Accessors
import org.apache.commons.exec.CommandLine
import org.apache.commons.lang3.ArrayUtils
import org.cobbzilla.util.collection.SingletonList

import java.io.ByteArrayInputStream
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.util.ArrayList

import org.cobbzilla.util.daemon.ZillaRuntime.empty
import org.cobbzilla.util.io.FileUtil.abs
import org.cobbzilla.util.string.StringUtil.UTF8cs

@Accessors(chain = true)
class Command {

    private var commandLine: CommandLine? = null
    private var input: String? = null
    private var rawInput: ByteArray? = null
    private var stdin: InputStream? = null
    private var dir: File? = null
    private var env: Map<String, String>? = null
    private var exitValues: MutableList<Int> = DEFAULT_EXIT_VALUES

    private var copyToStandard = false

    private var out: OutputStream? = null

    private var err: OutputStream? = null
    val inputStream: InputStream?
        get() {
            if (!hasInput()) return null
            if (stdin != null) return stdin
            return if (rawInput != null) ByteArrayInputStream(rawInput!!) else ByteArrayInputStream(input!!.toByteArray(UTF8cs))
        }

    constructor() {}

    fun hasOut(): Boolean {
        return out != null
    }

    fun hasErr(): Boolean {
        return err != null
    }

    constructor(commandLine: CommandLine) {
        this.commandLine = commandLine
    }

    constructor(command: String) : this(CommandLine.parse(command)) {}
    constructor(executable: File) : this(abs(executable)) {}

    fun hasDir(): Boolean {
        return dir != null
    }

    fun hasInput(): Boolean {
        return !empty(input) || !empty(rawInput) || stdin != null
    }

    fun getExitValues(): IntArray {
        return if (exitValues === DEFAULT_EXIT_VALUES)
            DEFAULT_EXIT_VALUES_INT
        else
            ArrayUtils.toPrimitive(exitValues.toTypedArray())
    }

    fun setExitValues(values: MutableList<Int>): Command {
        this.exitValues = values
        return this
    }

    fun setExitValues(values: IntArray): Command {
        exitValues = ArrayList(values.size)
        for (v in values) exitValues.add(v)
        return this
    }

    fun getCommandLine(): CommandLine? {
        return this.commandLine
    }

    fun getInput(): String? {
        return this.input
    }

    fun getRawInput(): ByteArray? {
        return this.rawInput
    }

    fun getStdin(): InputStream? {
        return this.stdin
    }

    fun getDir(): File? {
        return this.dir
    }

    fun getEnv(): Map<String, String>? {
        return this.env
    }

    fun isCopyToStandard(): Boolean {
        return this.copyToStandard
    }

    fun getOut(): OutputStream? {
        return this.out
    }

    fun getErr(): OutputStream? {
        return this.err
    }

    fun setCommandLine(commandLine: CommandLine): Command {
        this.commandLine = commandLine
        return this
    }

    fun setInput(input: String): Command {
        this.input = input
        return this
    }

    fun setRawInput(rawInput: ByteArray): Command {
        this.rawInput = rawInput
        return this
    }

    fun setStdin(stdin: InputStream): Command {
        this.stdin = stdin
        return this
    }

    fun setDir(dir: File): Command {
        this.dir = dir
        return this
    }

    fun setEnv(env: Map<String, String>): Command {
        this.env = env
        return this
    }

    fun setCopyToStandard(copyToStandard: Boolean): Command {
        this.copyToStandard = copyToStandard
        return this
    }

    fun setOut(out: OutputStream): Command {
        this.out = out
        return this
    }

    fun setErr(err: OutputStream): Command {
        this.err = err
        return this
    }

    companion object {

        val DEFAULT_EXIT_VALUES: MutableList<Int> = SingletonList(0)
        val DEFAULT_EXIT_VALUES_INT = intArrayOf(0)
    }
}
